packages/adapters-library/src/adapters/pendle/products/market/pendleMarketAdapter.ts

```diff
@@ -0,0 +1,305 @@
+    this.provider = provider
+    this.chainId = chainId
+    this.protocolId = protocolId
+    this.adaptersController = adaptersController
+    this.helpers = helpers
+  }
+
+  async unwrap({
+    blockNumber,
+    protocolTokenAddress,
+    tokenId,
+  }: UnwrapInput): Promise<UnwrapExchangeRate> {
+    const metadata = (await this.buildMetadata())[protocolTokenAddress]
+    const underlyingToken = metadata!.underlyingToken
+
+    const oracle = OraclePyYtLp__factory.connect(
+      '0x9a9fa8338dd5e5b2188006f1cd2ef26d921650c2',
+      this.provider,
+    )
+
+    let rate: bigint
+    switch (metadata!.type) {
+      case 'pt':
+        rate = await oracle
+          .getPtToSyRate(metadata!.marketAddress, 900, {
+            blockTag: blockNumber,
+          })
+          .catch((e) => {
+            return oracle.getYtToSyRate(metadata!.marketAddress, 1800, {
+              blockTag: blockNumber,
+            })
+          })
+
+        break
+      case 'yt':
+        rate = await oracle
+          .getYtToSyRate(metadata!.marketAddress, 900, {
+            blockTag: blockNumber,
+          })
+          .catch((e) => {
+            return oracle.getYtToSyRate(metadata!.marketAddress, 1800, {
+              blockTag: blockNumber,
+            })
+          })
+
+        break
+      case 'lp':
+        rate = await oracle
+          .getPtToSyRate(metadata!.marketAddress, 900, {
+            blockTag: blockNumber,
+          })
+          .catch((e) => {
+            return oracle.getYtToSyRate(metadata!.marketAddress, 1800, {
+              blockTag: blockNumber,
+            })
+          })
+
+        break
+      case 'sy': {
+        rate = 10n ** 18n
+
+        break
+      }
+
+      default:
+        throw new Error('Invalid metadata type')
+    }
+
+    const underlying = {
+      type: TokenType.Underlying,
+
+      underlyingRateRaw: rate,
+      ...underlyingToken,
+    }
+
+    return {
+      baseRate: 1,
+      type: TokenType.Protocol,
+      ...metadata!.protocolToken!,
+      tokens: [underlying],
+    }
+  }
+
+  getProtocolDetails(): ProtocolDetails {
+    return {
+      protocolId: this.protocolId,
+      name: 'Pendle',
+      description: 'Pendle Market adapter',
+      siteUrl: 'https://www.pendle.finance',
+      iconUrl: 'https://app.pendle.finance/favicon.ico',
+      positionType: PositionType.Supply,
+      chainId: this.chainId,
+      productId: this.productId,
+      assetDetails: {
+        type: AssetType.StandardErc20,
+      },
+    }
+  }
+
+  @CacheToFile({ fileKey: 'market' })
+  async buildMetadata(): Promise<Metadata> {
+    const resp = await fetchAllMarkets(this.chainId)
+
+    const metadata: Metadata = {}
+
+    resp.results.map((value) => {
+      const market = getAddress(value.address)
+
+      const pt: Erc20Metadata = {
+        address: getAddress(value.pt.address),
+        name: value.pt.name,
+        symbol: value.pt.symbol,
+        decimals: value.pt.decimals,
```

Comment:
`value.underlyingAsset` cannot be undefined according to their type, and none of the properties need `!`.

Reply:
yes, long story, was switching out tokens from their api, the shape of their metadata was not obvious, must have had a token there that was optional, good spot